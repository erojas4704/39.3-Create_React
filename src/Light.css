.light.on{
    color: black;
    background-color: yellow;
}

.light.off{
    background-color: black;
    color: white;
}

.light{
    width: 40px;
    height: 40px;
    border-radius: 50px;
}

.light > button{
	background: none;
	color: inherit;
	border: none;
	padding: 0;
	font: inherit;
	cursor: pointer;
	outline: inherit;
}