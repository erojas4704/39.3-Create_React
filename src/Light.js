import { useState } from "react";

const Light = ({ on,  onToggle, probabilityLightIsOn = 0.5 }) => {
    let [isOn, setIsOn] = useState( on !== undefined? on : probabilityLightIsOn > Math.random());

    const toggle = () => {
        setIsOn(!isOn);
        onToggle();
    }

    return (
        <span key={this} className="light">
            <button onClick={toggle}>{isOn ? 'Off' : 'On'}</button>
        </span>
    );
};

export default Light;