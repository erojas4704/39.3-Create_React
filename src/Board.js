import { useState } from "react";
import Light from "./Light";
import "./Board.css";

const generate2dArray = (width, height, probabilityLightIsOn) => {
    const arr = [];
    for (let x = 0; x < width; x++) {
        arr[x] = [];
        for (let y = 0; y < height; y++) {
            arr[x][y] = Math.random() > probabilityLightIsOn
        }
    }
    return arr;
}

const Board = ({
    width = 3,
    height = 3,
    probablityLightIsOn = 0.5
}) => {
    let [lights, setLights] = useState(generate2dArray(width, height, probablityLightIsOn));
    let [gameWon, setGameWon] = useState(false);

    function toggleAdjacentLights(origArray, x, y) {
        let alteredArray = [...origArray];
        alteredArray[x][y] = !alteredArray[x][y];
        if (x > 0) alteredArray[x - 1][y] = !alteredArray[x - 1][y];
        if (x < width - 1) alteredArray[x + 1][y] = !alteredArray[x + 1][y];
        if (y > 0) alteredArray[x][y - 1] = !alteredArray[x][y - 1];
        if (y < height - 1) alteredArray[x][y + 1] = !alteredArray[x][y + 1];
        return alteredArray;
    }

    function checkGameWon(array) {
        for (let x = 0; x < width; x++) {
            for (let y = 0; y < height; y++) {
                if (array[x][y] === true) {
                    return false;
                }
            }
        }
        return true;
    }


    const onToggle = (x, y, on) => {
        //Toggle adjacent lights
        let newLights = toggleAdjacentLights(lights, x, y);
        setLights(newLights);
        if (checkGameWon(newLights)) {
            setGameWon(true);
            console.log(`Game won!`);
        };
    };


    let grid = [];
    for (let y = 0; y < height; y++) {
        grid.push([]);
        for (let x = 0; x < width; x++) {
            grid[y].push(
                <Light
                    key={`${x}${y}`}
                    onToggle={onToggle}
                    on={lights[x][y]}
                    x={x}
                    y={y}
                />
            );
        }
    }

    return (
        <div className="board">
            <h3>Lights Out</h3>
            {gameWon ? (<div>You win forever</div>) : grid.map((row, y) => {
                return (
                    <div key={y}>
                        {row}
                    </div>
                )
            })}
        </div>
    );
}

export default Board;
