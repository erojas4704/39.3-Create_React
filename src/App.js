import logo from './logo.svg';
import './App.css';
import Board from './Board';

function App() {
  return (
    <>
      <Board width={3} height={3} probablityLightIsOn={0.5} />
    </>
  );
}

export default App;
