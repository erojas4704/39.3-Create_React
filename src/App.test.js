import { render, screen } from '@testing-library/react';
import App from './App';
import Board from './Board';
import Light from './Light';

test('renders', () => {
  render(<App />);
});

test('renders cells properly', () => {
  render(<Light />);
});

test('renders board properly', () => {
  render(<Board />);
});

test('compares against snapshot', () => {
  const { asFragment } = render(<App />);
  expect(asFragment).toMatchSnapshot();
});