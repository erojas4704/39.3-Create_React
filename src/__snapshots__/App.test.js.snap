// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compares against snapshot 1`] = `[Function]`;
